src/Compilers/Core/Portable/Syntax/SyntaxDiffer.cs

```diff
@@ -505,7 +505,17 @@ private int GetSimilarity(in SyntaxNodeOrToken node1, in SyntaxNodeOrToken node2
 
         private static bool AreIdentical(in SyntaxNodeOrToken node1, in SyntaxNodeOrToken node2)
         {
-            return node1.UnderlyingNode == node2.UnderlyingNode;
+            if (node1.UnderlyingNode == null && node2.UnderlyingNode == null)
+            {
+                return true;
+            }
+
+            if (node1.UnderlyingNode == null && node2.UnderlyingNode != null)
+            {
+                return false;
+            }
+
+            return node1.UnderlyingNode!.IsEquivalentTo(node2.UnderlyingNode);
```

Comment:
this seems wrong.  the AreIdentical was very likely intending to ask the exact question of if the green nodes were identical.  if that's not appropriate, it's likely we need to fixup the caller of this to adopt whatever semantics we need.

Reply:
Ahh on second glance, I don't know if `IsEquivalentTo` will even work here. `==` tests object ref comparison and `IsEquivalentTo` tests to see if the [objects are structurally equivalent](https://github.com/dotnet/roslyn/blob/9c10b941e96760e0c986c8f2afb0d5b4c34e67e3/src/Compilers/Core/Portable/Syntax/GreenNode.cs#L837). However, I think in order to determine if the objects are equal we need to test for syntactic equivalence too, right? For example, I think two nodes of the same type e.g. `A` and `B` would return true in `IsEquivalentTo` since they're structurally the same, even though syntactically they're different.
@CyrusNajmabadi would you be aware of any helper method that will test for syntactic equivalence of two nodes?